static/js/drafts.js

```diff
@@ -0,0 +1,227 @@
+        var drafts = get();
+
+        if (drafts[id]) {
+            draft.updatedAt = getTimestamp();
+            drafts[id] = draft;
+            save(drafts);
+        }
+    };
+
+    exports.deleteDraft = function (id) {
+        var drafts = get();
+
+        delete drafts[id];
+        save(drafts);
+    };
+
+    return exports;
+}());
+
+exports.draft_model = draft_model;
+
+exports.update_draft = function () {
+    var draft = compose.snapshot_message();
+    var draft_id = $("#new_message_content").data("draft-id");
+
+    if (draft_id !== undefined) {
+        if (draft !== undefined) {
+            draft_model.editDraft(draft_id, draft);
+        } else {
+            draft_model.deleteDraft(draft_id);
+        }
+    } else {
+        if (draft !== undefined) {
+            var new_draft_id = draft_model.addDraft(draft);
+            $("#new_message_content").data("draft-id", new_draft_id);
+        }
+    }
+};
+
+exports.delete_draft_after_send = function () {
+    var draft_id = $("#new_message_content").data("draft-id");
+    if (draft_id) {
+        draft_model.deleteDraft(draft_id);
+    }
+    $("#new_message_content").removeData("draft-id");
+};
+
+exports.restore_draft = function (draft_id) {
+    var draft = draft_model.getDraft(draft_id);
+    if (!draft) {
+        return;
+    }
+
+    var draft_copy = _.extend({}, draft);
+    if ((draft_copy.type === "stream" &&
+         draft_copy.stream.length > 0 &&
+             draft_copy.subject.length > 0) ||
+                 (draft_copy.type === "private" &&
+                  draft_copy.reply_to.length > 0)) {
+        draft_copy = _.extend({replying_to_message: draft_copy},
+                              draft_copy);
+    }
+
+    exports.close();
+    compose_fade.clear_compose();
+    if (draft.type === "stream" && draft.stream === "") {
+      draft_copy.subject = "";
+    }
+    compose.start(draft_copy.type, draft_copy);
+    $("#new_message_content").data("draft-id", draft_id);
+};
+
+exports.setup_page = function (callback) {
+    function setup_event_handlers() {
+        window.addEventListener("beforeunload", function () {
+            exports.update_draft();
+        });
+
+        $("#new_message_content").focusout(exports.update_draft);
+
+        $(".draft_controls .restore-draft").on("click", function (e) {
+            e.stopPropagation();
+
+            var draft_row = $(this).closest(".draft-row");
+            var draft_id = draft_row.data("draft-id");
+            exports.restore_draft(draft_id);
+        });
+
+        $(".draft_controls .delete-draft").on("click", function () {
+            var draft_row = $(this).closest(".draft-row");
+            var draft_id = draft_row.data("draft-id");
+
+            exports.draft_model.deleteDraft(draft_id);
+            draft_row.remove();
+
+            if ($("#drafts_table .draft-row").length === 0) {
+                $('#drafts_table .no-drafts').show();
+            }
+        });
+    }
```

Comment:
At least the `focusout` and `beforeunload` parts of this should be run once on page load, not the first time the drafts UI launches.  The current implementation is buggy in that it doesn't track drafts created before the user first opens the drafts UI.

You probably just want to put this code inside a `$(function () {`